[BE] feat: 헤더 존재 여부 검증

donghoony opened the pull request:
<!-- 제목: [BE/FE/All] (기능 등) -->
<!-- 아래에 이슈 번호를 매겨주세요 -->

- resolves #205 

---

### 🚀 어떤 기능을 구현했나요 ?
- Header에 원하는 필드가 존재하지 않으면, request.getHeader(name)은 null을 리턴합니다. 해당 값이 서비스와 도메인단까지 내려가 실제 쿼리로 이어집니다. 이는 의도한 것이 아니므로, 만약 존재하지 않으면 예외를 발생하게 합니다.


### 🔥 어떻게 해결했나요 ?
- 커스텀 어노테이션 및 밸리데이터를 추가했습니다.
- `Advice`에서 예외를 적절히 처리하도록 변경하고, 메시지가 클라이언트에 전달되도록 합니다.

### 📝 어떤 부분에 집중해서 리뷰해야 할까요?
- 클라이언트까지 메시지를 보내기 위해서 `HandlerMethodValidationException`을 따로 뽑았습니다. 이 방법이 최선인지 확인해줘요!

### 할 말
- 예외 내부의 모든 곳에 로깅이 필요해 보여요. 어떤 메서드에서 걸리는지 확인하는 게 어려웠습니다.

donghoony committed:
fix: 인터페이스와 구현체 어노테이션 일치

donghoony committed:
feat: 헤더 검사 어노테이션

donghoony committed:
feat: 헤더 밸리데이터

donghoony committed:
feat: 컨트롤러에 헤더 검사

donghoony committed:
feat: 헤더 검증 메시지 클라이언트에 전달

donghoony commented on backend/src/main/java/reviewme/review/controller/validator/ContainsHeaderName.java:
```diff
@@ -0,0 +1,26 @@
+    String value() default "";
+
+    @AliasFor("value")
+    String headerName() default "";
```
`@AliasFor`는 어노테이션 내부에 `headerName=`을 적지 않고 바로 적용할 수 있도록 합니다. 기본적으로 아무것도 적지 않으면 `value`와 같은데, 이를 매핑한 것이라고 보면 돼요.

donghoony commented on backend/src/main/java/reviewme/review/controller/ReviewController.java:
```diff
@@ -45,7 +48,10 @@ public ResponseEntity<ReviewSetupResponse> findReviewCreationSetup(@RequestParam
-    public ResponseEntity<ReviewDetailResponse> findReceivedReviewDetail(@PathVariable long id, HttpServletRequest request) {
+    public ResponseEntity<ReviewDetailResponse> findReceivedReviewDetail(
+            @PathVariable long id,
+            @ContainsHeaderName(GROUP_ACCESS_CODE_HEADER) HttpServletRequest request
```
범용적으로 사용할 것을 대비해 `@ContainsHeaderName`이라고 지었어요. 만약 단순 헤더 이름이 `GroupAccessCode`였다면 그렇게도 할 수 있었겠지만.. 이 부분은 따로 빼서 값을 넣어주는 식으로 하는 게 나아 보였습니다.

donghoony committed:
fix: 누락된 `@Valid` 어노테이션 추가

donghoony committed:
refactor: 요청이 null인 경우 핸들링

donghoony committed:
chore: 테스트 이름 간결하게 변경

donghoony commented on backend/src/main/java/reviewme/review/controller/validator/ContainsHeaderValidator.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public boolean isValid(HttpServletRequest request, ConstraintValidatorContext context) {
+        if (request == null) {
+            return false;
+        }
```
request에 null이 오는 상황은 통상적으로 존재하지 않지만, 테스트(직접 메서드 호출) / Request Mocking 등의 상황에서 충분히 발생할 수 있는 상황이라 원천봉쇄하는 게 낫다고 생각했어요.

donghoony committed:
chore: 소문자 컨벤션

hyeonji1220 commented on backend/src/main/java/reviewme/global/GlobalExceptionHandler.java:
```diff
@@ -103,4 +103,14 @@ public ProblemDetail handleMethodArgumentNotValid(MethodArgumentNotValidExceptio
         return problemDetail;
     }
+
+    @ExceptionHandler(HandlerMethodValidationException.class)
+    public ProblemDetail handleHandlerMethodValidationException(HandlerMethodValidationException ex) {
+        String message = ex.getAllErrors()
+                .stream()
+                .map(MessageSourceResolvable::getDefaultMessage)
+                .toList()
+                .get(0);
+        return ProblemDetail.forStatusAndDetail(HttpStatus.BAD_REQUEST, message);
+    }
```
권한에 해당하는 헤더 정보가 없이 온 경우에 대해서 HandlerMethodValidationException를 내려주는 게 적절한지 궁금합니다!
권한에 해당하는 헤더 정보가 없는 경우는 인증의 문제 느낌이 있기도 해서요!

donghoony replied:
원복했습니다 👍🏻

hyeonji1220 commented on backend/src/main/java/reviewme/review/controller/ReviewController.java:
```diff
@@ -45,7 +48,10 @@ public ResponseEntity<ReviewSetupResponse> findReviewCreationSetup(@RequestParam
-    public ResponseEntity<ReviewDetailResponse> findReceivedReviewDetail(@PathVariable long id, HttpServletRequest request) {
+    public ResponseEntity<ReviewDetailResponse> findReceivedReviewDetail(
+            @PathVariable long id,
+            @Valid @ContainsHeaderName(GROUP_ACCESS_CODE_HEADER) HttpServletRequest request
```
사실 어노테이션을 만들고, 그 안에 validator를 연관시켜서 검증을 하는 방식을 처음봐서 이해중입니다만..😁
똑같이 어노테이션을 만들어서 ArgumentResolver가 그걸 인식해서 검증하는 방식이 처음에 떠올랐는데 
그와 비교했을 때 해당 방식의 장점이 궁금해요!

donghoony replied:
확실히 너무 빠졌네요, Resolver에서 `@HeaderProperty`를 찾도록 변경하는 게 나아 보입니다 !

hyeonji1220 reviewed (changes requested):
어노테이션을 만들고.. @AliasFor를 사용하고.. 신기방기합니다🤩🤩🤩

donghoony committed:
feat: ArgumentResolver를 활용한 헤더 검증

donghoony committed:
feat: 공용 패키지로 이동 및 범용적으로 사용할 수 있도록 수정

donghoony committed:
chore: 사용하지 않는 import 제거

donghoony committed:
feat: ArgumentResolver 적용, Validator 삭제

donghoony committed:
chore: 사용하지 않는 커스텀 예외 원복

donghoony committed:
refactor: 헤더 존재하지 않는 경우 메시지 수정

donghoony committed:
chore: resolve conflict

Merge remote-tracking branch 'refs/remotes/origin/develop' into be/feature/205-validate-header-name

# Conflicts:
#	backend/src/main/java/reviewme/review/controller/ReviewController.java

hyeonji1220 commented on backend/src/main/java/reviewme/review/controller/ReviewController.java:
```diff
@@ -45,9 +46,9 @@ public ResponseEntity<ReviewSetupResponse> findReviewCreationSetup(@RequestParam
-    public ResponseEntity<ReviewDetailResponse> findReceivedReviewDetail(@PathVariable long id,
-                                                                         HttpServletRequest request) {
-        String groupAccessCode = request.getHeader(GROUP_ACCESS_CODE_HEADER);
+    public ResponseEntity<ReviewDetailResponse> findReceivedReviewDetail(
+            @PathVariable long id,
+            @HeaderProperty(GROUP_ACCESS_CODE_HEADER) String groupAccessCode) {
```
이렇게 컨트롤러에서 헤더 이름을 넣어주고 어노테이션을 만들어서 value로 관리하니까 컨트롤러 한 곳만 headerName을 알고있게되어 좋군요!

근데 다른 컨트롤러에서도 이 헤더를 확인해야하는 경우가 생긴다면 별도의 클래스 등으로 빼야할 수도 있겠다는 생각이 듭니다!

hyeonji1220 commented on backend/src/test/java/reviewme/global/HeaderPropertyArgumentResolverTest.java:
오호 이렇게 ArgumentResolver를 테스트할 때 MethodParameter와 HeaderProperty를 mock 처리해서 테스트를 한다라... 저도 따라 쳐보겠습니다👋

hyeonji1220 reviewed (approved):
발벗고 코드 개선에 힘쓰는 아루... 저 자세 배우고만다...🦶

nayonsoso commented on backend/src/main/java/reviewme/global/HeaderPropertyArgumentResolver.java:
```diff
@@ -0,0 +1,32 @@
+        String headerName = parameterAnnotation.headerName();
+        String headerProperty = request.getHeader(headerName);
+
+        if (headerProperty == null) {
+            throw new MissingHeaderPropertyException(headerName);
+        }
+        return headerProperty;
+    }
```
은제 바도 멋진 코드.. 🌹

지정한 이름의 헤더가 Blank 일 때는 예외를 던지지 않고, null 일 때만 예외처리를 하게 한 것에는 어떤 의도가 있으신지 궁금헙니다...

donghoony replied:
문제가 되는 부분이 헤더가 존재하지 않을 때 `null`이 쿼리까지 들어가는 거라서, 이를 방지하고자 했습니다.

nayonsoso commented on backend/src/main/java/reviewme/review/controller/ReviewApi.java:
```diff
@@ -19,7 +20,7 @@ public interface ReviewApi {
             @ApiResponse(responseCode = "201", description = "리뷰 등록 요청 성공")
     })
-    ResponseEntity<Void> createReview(@RequestBody CreateReviewRequest request);
+    ResponseEntity<Void> createReview(@Valid @RequestBody CreateReviewRequest request);
```
왜 api 문서에도 valid 를 다신건지.. 궁금헙니다.. <--@

donghoony replied:
이거 없으니까 안 돌아가던데요 !?

donghoony replied:
뭐지 없어도 되었네요..? 🤔 Validator 추가할 때 어딘가 꼬였나봐요, 다시 돌려놓겠습니다 👍🏻

nayonsoso commented on backend/src/test/java/reviewme/global/HeaderPropertyArgumentResolverTest.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Test
+    void 검증값이_헤더에_존재하면_값을_반환한다() {
+        // given
+        NativeWebRequest request = mock(NativeWebRequest.class);
+        MockHttpServletRequest mockRequest = new MockHttpServletRequest();
+        mockRequest.addHeader("test", "1234");
+        given(request.getNativeRequest()).willReturn(mockRequest);
+        given(headerProperty.headerName()).willReturn("test");
+
+        // when
+        String actual = resolver.resolveArgument(parameter, null, request, null);
+
+        // then
+        assertThat(actual).isEqualTo("1234");
+    }
```
```suggestion
        // given
        String headerName = "test";
        String headerValue = "1234";
        NativeWebRequest request = mock(NativeWebRequest.class);
        MockHttpServletRequest mockRequest = (new MockHttpServletRequest());
        mockRequest.addHeader(headerName, headerValue);
        given(request.getNativeRequest()).willReturn(mockRequest);
        given(headerProperty.headerName()).willReturn(headerName);

        // when
        String actual = resolver.resolveArgument(parameter, null, request, null);

        // then
        assertThat(actual).isEqualTo(headerValue);
```

테스트 코드에서는 변수명을 통해 의미 전달을 하는게 좋을 것 같읍니다..

donghoony replied:
반영할게요~

nayonsoso reviewed (commented):
한사랑 산악회 컨셉으로 리뷰 달아봤습니다 ㅎㅎ
argumentResolver 를 테스트하는 방법을 보고 진짜 깜짝 놀랐습니다!!
저렇게 테스트 코드를 짤 수도 있군요🤩 배워갑니당~~

![image](https://github.com/user-attachments/assets/1f36b684-4d5a-4691-9a72-59e31b7d9e45)

donghoony committed:
chore: 불필요한 `@Valid` 어노테이션 삭제

donghoony committed:
refactor: 변수 추출

Co-authored-by: Yeongseo Na <panda0329@naver.com>

nayonsoso commented on backend/src/main/java/reviewme/review/controller/ReviewApi.java:
```diff
@@ -20,7 +19,7 @@ public interface ReviewApi {
             @ApiResponse(responseCode = "201", description = "리뷰 등록 요청 성공")
     })
-    ResponseEntity<Void> createReview(@Valid @RequestBody CreateReviewRequest request);
+    ResponseEntity<Void> createReview(@RequestBody CreateReviewRequest request);
```
👍

donghoony committed:
fix: 컴파일 에러 해결

nayonsoso commented on backend/src/test/java/reviewme/global/HeaderPropertyArgumentResolverTest.java:
```diff
@@ -52,4 +52,5 @@ void setUp() {
 
         // then
         assertThat(actual).isEqualTo(headerValue);
+    }
```
👍

nayonsoso reviewed (approved):
은제바도 머찐 코두입니다

Kimprodp reviewed (approved):
고생햇슴니다. swagger 사용을 위한 @RequestHeader 는 같이 좀 더 고민해봅시당~
